src/BotPlutusInterface/TimeSlot.hs

```diff
@@ -0,0 +1,174 @@
+import Ledger (Extended (Finite, NegInf, PosInf), Interval (Interval), LowerBound (LowerBound), UpperBound (UpperBound))
+import Ledger qualified
+import Ouroboros.Consensus.HardFork.History qualified as Consensus
+import Ouroboros.Consensus.HardFork.History.Qry qualified as HF
+import System.Environment (getEnv)
+import Prelude
+
+-- | Error returned by the functions of this module
+data TimeSlotConversionError
+  = TimeSlotConversionError !Text
+  deriving stock (Show)
+
+-- Analogous to `slotToBeginPOSIXTime` from plutus-ledger
+
+-- | Convert `Slot` to `POSIXTime`.
+slotToPOSIXTimeIO :: PABConfig -> Ledger.Slot -> IO (Either TimeSlotConversionError Ledger.POSIXTime)
+slotToPOSIXTimeIO pabConf lSlot = runEitherT $ do
+  nodeInfo <- liftIO $ mkNodeInfo pabConf
+  eraHsitory <- newET (queryEraHistory nodeInfo)
+  sysStart <- newET $ querySystemStart nodeInfo
+  let epochInfo = toLedgerEpochInfo eraHsitory
+      pparams =
+        CAPI.toLedgerPParams
+          CAPI.ShelleyBasedEraAlonzo
+          (pcProtocolParams pabConf)
+
+  firstEitherT toError . hoistEither $
+    slotToPOSIXTime pparams epochInfo sysStart (toSlotNo lSlot)
+
+-- | Convert `POSIXTime` to `Slot`.
+
+-- Analogous to `posixTimeToEnclosingSlot` from plutus-ledger
+posixTimeToSlotIO ::
+  PABConfig ->
+  Ledger.POSIXTime ->
+  IO (Either TimeSlotConversionError Ledger.Slot)
+posixTimeToSlotIO pabConf pTime = runEitherT $ do
+  nodeInfo <- liftIO $ mkNodeInfo pabConf
+  eraHist <- newET (queryEraHistory nodeInfo)
+  sysStart <- newET $ querySystemStart nodeInfo
+  firstEitherT toError . hoistEither $
+    posixTimeToSlot sysStart eraHist pTime
+
+{- | Convert a `POSIXTimeRange` to `SlotRange`.
+ Gives the biggest slot range that is entirely contained by the given time range.
+-}
+
+-- Analogous to `posixTimeRangeToContainedSlotRange` from plutus-ledger
+posixTimeRangeToContainedSlotRangeIO ::
+  PABConfig ->
+  Ledger.POSIXTimeRange ->
+  IO (Either TimeSlotConversionError Ledger.SlotRange)
+posixTimeRangeToContainedSlotRangeIO
+  pabConf
+  ptr@(Interval (LowerBound start startIncl) (UpperBound end endIncl)) = runEitherT $ do
+    -- getting required info from node
+    nodeInfo <- liftIO $ mkNodeInfo pabConf
+    sysStart <- newET $ querySystemStart nodeInfo
+    eraHsitory <- newET (queryEraHistory nodeInfo)
+    let epochInfo = toLedgerEpochInfo eraHsitory
+        pparams =
+          CAPI.toLedgerPParams
+            CAPI.ShelleyBasedEraAlonzo
+            (pcProtocolParams pabConf)
+
+    let extTimeToExtSlot = convertExtended sysStart eraHsitory
+        getClosure = getExtClosure pparams epochInfo sysStart
+
+    -- conversions
+    startSlot <- extTimeToExtSlot start
+    startSlotClosure <- getClosure startSlot startIncl
+    endSlot <- extTimeToExtSlot end
+    endSlotClosure <- getClosure endSlot endIncl
+    let lowerB = LowerBound startSlot startSlotClosure
+        upperB = UpperBound endSlot endSlotClosure
+    pure $ Interval lowerB upperB
+    where
```

Comment:
These are a litttle involved, lets get some type signatures on them

Reply:
Added signatures there as well and clarification comments.